### PR TITLE
Fixing typing for ListenerMixin.listener

### PR DESCRIPTION
**Pull request by aericson**
### Description

This PR addresses #2375. The typing of that method is a bit complex because it has two different code paths which result in different typing of parameters and return.

I've added two overloads so that we get better type checking.

### Alternative solutions

I've also considered another way to go around this by separating `listener` into two methods which should reduce complexity but in this PR I only addressed the typing.

Here is how the alternative solution could look like:
https://gist.github.com/aericson/d36514eb45b9ec1f65b8ee51fdf5c08f

### CODE REVIEW

File: sanic/models/handler_types.py [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
 from typing import Any, Callable, Coroutine, Optional, TypeVar, Union
 
+import sanic
+
 from sanic.request import Request
 from sanic.response import BaseHTTPResponse, HTTPResponse
 
 
-Sanic = TypeVar("Sanic")
+Sanic = TypeVar("Sanic", bound="sanic.Sanic")
```
Comment (aericson) [inline sanic/models/handler_types.py:10]: I'm not super sure about this change. Not sure we should do it, and maybe outside the scope of the PR.

Comment (ahopkins) [reply sanic/models/handler_types.py:10]: This is okay

**Comment by ahopkins** [comment]
> I've also considered another way to go around this by separating `listener` into two methods which should reduce complexity but in this PR I only addressed the typing.

Let's consider that in a separate PR.

File: sanic/mixins/listeners.py [hunk 2]
```diff
@@ -59,6 +82,8 @@ def register_listener(listener, event):
             return listener
 
         if callable(listener_or_event):
+            if TYPE_CHECKING:
+                assert event_or_none is not None
```
Comment (ahopkins) [inline sanic/mixins/listeners.py:86]: Avoid `assert` in production code. There is another instance in the codebase where we have this, but we should not proliferate the issue. Use `raise` instead.

File: sanic/mixins/listeners.py [hunk 3]
```diff
@@ -26,12 +26,33 @@ def __init__(self, *args, **kwargs) -> None:
     def _apply_listener(self, listener: FutureListener):
         raise NotImplementedError  # noqa
 
+    @overload
+    def listener(
+        self,
+        listener_or_event: ListenerType[Sanic],
+        event_or_none: str,
+        apply: bool = True,
```
Comment (ahopkins) [inline sanic/mixins/listeners.py:34]: Do not need a default here.

File: sanic/mixins/listeners.py [hunk 4]
```diff
@@ -26,12 +26,33 @@ def __init__(self, *args, **kwargs) -> None:
...
+    ) -> ListenerType[Sanic]:
+        ...
+
+    @overload
+    def listener(
+        self,
+        listener_or_event: str,
+        event_or_none: None = None,
+        apply: bool = True,
```
Comment (ahopkins) [inline sanic/mixins/listeners.py:43]: Can remove these defaults

File: sanic/mixins/listeners.py (hunk 3 above)
Comment (aericson) [reply sanic/mixins/listeners.py:34]: won't this make the typing invalid? I can't get the type checking to work as intended without the default. Removing them would make them required.

Comment (aericson) [reply sanic/mixins/listeners.py:34]: Oh, nice. Just found out one can use an ellipsis (`...`). Thanks for teaching me that 😄

**Review by ahopkins — APPROVED** [review]

**Comment by ahopkins** [comment]
> ### Alternative solutions
> 
> I've also considered another way to go around this by separating `listener` into two methods which should reduce complexity but in this PR I only addressed the typing.

No rush on this, but do you think you can put this alternative version together as a PR for the June release?

**Comment by aericson** [comment]
> > ### Alternative solutions
> > 
> > I've also considered another way to go around this by separating `listener` into two methods which should reduce complexity but in this PR I only addressed the typing.
> 
> No rush on this, but do you think you can put this alternative version together as a PR for the June release?

Yep, I can put some time into that. Are you thinking on us closing this PR and having a new one with that? Just wondering if I should do on top of this changes or from a clean branch.

**Pull request merged**